agent/connect/ca/provider_vault.go

```diff
@@ -173,6 +179,11 @@ func (v *VaultProvider) renewToken(ctx context.Context, watcher *vaultapi.Lifeti
 			return
 
 		case err := <-watcher.DoneCh():
+			// In the event we fail to login to Vault or our token is no longer valid we can overwhelm a Vault
+			// instance with rate limit configured. We would make these requests to Vault as fast as we possibly
+			// could and start causing all client's to receive 429 response codes. To mitigate that we're sleeping 1
+			// second or less before restarting the LifetimeWatcher. Once we can upgrade to
+			// github.com/hashicorp/vault/api@v1.1.1 or later the LifetimeWatcher _should_ perform that backoff for us.
```

Comment:
What was the change on Vault side? I see backoff in their code from over a year ago.

If we upgrade to 1.1.1, do we still need to sleep here? If so we could maybe use an exponential backoff too - it's still a bit gross to be making on average 2 failing auth calls per second forever if we have a bad token. I'd probably use our `lib.Retry` for it in `/lib`.

Finally, is there something stopping us from upgrading to api@1.1.1 now?

Reply:
The change of the Vault side _was_ the addition of the exponential backoff from over a year ago. If we upgrade to 1.1.1 we do not need to sleep here because the watcher should get an error, retry with the exponential backoff, then exit. The Vault code is currently returning attempting to renew, returning quickly and we are immediately restarting it causing another renew attempt.

Upgrading to api@1.1.1 upgrades some other dependencies including protobuf, the aws-sdk, and genproto. I'm not super familiar with what those would have affect or what our normal process is for upgrading our dependencies so I put the sleep here instead. The exponential retry in the Vault code would cap out at 1 second so we'd still have 1 failing call per second forever (or until consul is reconfigured).

I can switch to lib.Retry fairly easily or upgrade to the newer vault/api version.

Reply:
Hmm yeah the protobuf thing probably means we can't until we are done moving off gogo but I'll defer to the Core team to work out if that's true!

I think this is OK for a temporary solution but maybe we could:
 - make it `lib.RandomJitter(4 *time.Second) + time.Second` so it's 1-5 seconds rather than 0-1. (or lib.Retry which is not too hard?) either seems fine
 - Add `TODO` comments here and maybe link this PR in the issue we are tracking protobuf upgrades so we don't forget to come back and upgrade?

I think it's worth trying to merge and ship this fix as soon as possible as it could affect production Vault stability.

@hashicorp/consul-core, what do you think?

Reply:
I've switched from RandomJitter to lib.Retry. I've also created a separate issue to upgrade to a newer version of the vault/api and linked it here and to https://github.com/hashicorp/consul/issues/10471